micrometer-test/src/main/java/io/micrometer/core/instrument/InstrumentationTimingVerificationTests.java

```diff
@@ -0,0 +1,108 @@
+import io.micrometer.observation.tck.TestObservationRegistryAssert;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@ExtendWith(InstrumentationVerificationTests.AfterBeforeParameterResolver.class)
+abstract class InstrumentationTimingVerificationTests extends InstrumentationVerificationTests {
+
+    /**
+     * A default is provided that should be preferred by new instrumentations. Existing
+     * instrumentations that use a different value to maintain backwards compatibility may
+     * override this method to run tests with a different name used in assertions.
+     * @return name of the Timer meter produced from the timing instrumentation under test
+     */
+    protected abstract String timerName();
+
+    /**
+     * If a {@link ObservationDocumentation} is provided the tests run will check that the
+     * produced instrumentation matches the given {@link ObservationDocumentation}.
+     * @return the documented observation to compare results against, or null to do
+     * nothing
+     */
+    @Nullable
+    protected ObservationDocumentation observationDocumentation() {
+        return null;
+    }
+
+    @AfterEach
+    void verifyObservationDocumentation(TestType testType) {
+        ObservationDocumentation observationDocumentation = observationDocumentation();
+        if (observationDocumentation == null) {
+            return;
+        }
+
+        Timer timer = getRegistry().get(timerName()).timer();
+        Set<String> requiredDocumentedLowCardinalityKeys = getRequiredLowCardinalityKeyNames(observationDocumentation);
+        Set<String> requiredTagKeys = new HashSet<>(requiredDocumentedLowCardinalityKeys);
+        if (testType == TestType.METRICS_VIA_OBSERVATIONS_WITH_METRICS_HANDLER) {
+            requiredTagKeys.add("error");
+        }
+        Set<String> allDocumentedLowCardinalityKeys = getLowCardinalityKeyNames(observationDocumentation);
+        Set<String> allPossibleTagKeys = new HashSet<>(allDocumentedLowCardinalityKeys);
+        if (testType == TestType.METRICS_VIA_OBSERVATIONS_WITH_METRICS_HANDLER) {
+            allPossibleTagKeys.add("error");
+        }
+
+        // must have all required tag keys
+        assertThat(timer.getId().getTags()).extracting(Tag::getKey).containsAll(requiredTagKeys);
+        // must not contain tag keys that aren't documented
+        assertThat(timer.getId().getTags()).extracting(Tag::getKey).isSubsetOf(allPossibleTagKeys);
+
+        if (testType == TestType.METRICS_VIA_OBSERVATIONS_WITH_METRICS_HANDLER) {
+            if (observationDocumentation.getDefaultConvention() == null) {
+                TestObservationRegistryAssert.assertThat(getObservationRegistry())
```

Comment:
I wonder if all of that couldn't be hidden inside a new assertion in the `TestObservationRegistryAssert`

Reply:
Yes, probably. We can take a look after RC1 I think, as no users are asking for such an assertion now and there's other high priority stuff to work on.